src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -21,11 +22,588 @@
+
+		/*
+		 * If we found a node in the graph we only need to add referred node to
+		 * the adjacency  list of that node.
+		 */
+		if (referringFound)
+		{
+			/*
+			 * If referred node is already in the adjacency list of referred node, do nothing.
+			 */
+			if (referringNode->adjacencyList == NIL || !list_member(
+					referringNode->adjacencyList, referredNode))
+			{
+				/*
+				 * If referred node also exists, add it to the adjacency list
+				 * and continue.
+				 */
+				if (referredFound)
+				{
+					referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+														   referredNode);
+				}
+				else
+				{
+					/*
+					 * We need to setup the node and add it to the adjacency list
+					 * of referring node.
+					 */
+					referredNode->adjacencyList = NIL;
+					referredNode->index = curIndex++;
+					referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+														   referredNode);
+				}
+			}
+			else
+			{
+				continue;
+			}
+		}
+		else
+		{
+			/*
+			 * If referring node is not exist in the graph, set its remaining parameters.
+			 */
+			referringNode->adjacencyList = NIL;
+			referringNode->index = curIndex++;
+			if (referredFound)
+			{
+				referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+													   referredNode);
+			}
+			else
+			{
+				/*
+				 * We need to setup the node and add it to the adjacency list
+				 * of referring node.
+				 */
+				referredNode->adjacencyList = NIL;
+				referredNode->index = curIndex++;
+				referringNode->adjacencyList = lappend(referringNode->adjacencyList,
+													   referredNode);
+			}
+		}
+	}
+
+
+	frelGraph->nodeCount = curIndex;
+
+	if (frelGraph->nodeCount > 0)
+	{
+		/*
+		 * Initialize index to oid mapping, that is necessary for accessing node
+		 * elements in O(1) time.
+		 */
+		InitializeIndexToOidMapping();
+
+		/*
+		 * Transitivity matrix will be used to find affected and affecting relations
+		 * for foreign key relation graph.
+		 */
+		CreateTransitivityMatrix();
+	}
+
+	frelGraph->validGraph = true;
+	systable_endscan(fkeyScan);
+	heap_close(fkeyRel, AccessShareLock);
+}
+
+
+/*
+ * InitializeIndexToOidMapping set the element of indexToOidArray element of
+ * frelGraph. You need to be make sure that all of the nodes have been added
+ * to the frelGraph and nodeCount has been set before.
+ */
+static void
+InitializeIndexToOidMapping()
+{
+	HASH_SEQ_STATUS status;
+	FRelNode *frelNode = NULL;
+
+	frelGraph->indexToOidArray = (Oid *) palloc(sizeof(Oid) * frelGraph->nodeCount);
+
+	hash_seq_init(&status, frelGraph->nodeMap);
+	while ((frelNode = (FRelNode *) hash_seq_search(&status)) != 0)
+	{
+		uint32 relationIndex = frelNode->index;
+		frelGraph->indexToOidArray[relationIndex] = frelNode->relationId;
+	}
+}
+
+
+/*
+ * CreateTransitiveClosure creates the transitive closure matrix for the given
+ * graph. It uses Floyd Warshall algorithm for obtaining the transitivity graph.
```

Comment:
If this is _just_ to calculate a transitive closure (also, this is a _connectivity matrix_, not a _transitivity graph_), then the algorithm is just called _Warshall's Algorithm_. _Floyd-Warshall_ is for calculating shortest paths in a weighted graph (not what we're doing here).